# BIGCAAT
BIGDAWG Integrated GFE Customization and Amino Acid Testing (BIGCAAT) - software for standardizing and analyzing
HLA disease-associated motifs at a gene feature and amino acid level 

This repository contains all files and code relevant to Livia Tran’s Master’s Final Report. 

Multiple Sclerosis (MS) affects 2.3 million people globally by inducing persistent inflammation of the central nervous system. Prior to single nucleotide polymorphism (SNP) and genotype wide association studies (GWAS) technologies, research conducted via tissue typing showed strong correlation of MS with alleles of the Human Leukocyte Antigen (HLA) loci, a set of highly polymorphic genes located in the Major Histocompatibility Complex (MHC). HLA has been a gene of interest due to its strong association with disease and the immune system. However, due to HLA’s polymorphic nature, previous genotyping methods, such as Sanger sequencing, were unable to provide full gene sequence information, and only provided sequence data for a few key exons. Next generation sequencing (NGS) has made it possible to procure full gene sequences of most HLA alleles at high throughput, leading to better understanding of HLA’s impact on disease and healthcare. 


Files and functions are organized into Module 1 and Module 2.

Module 1 allows users to convert BIGDAWG formatted data to its Gene Feature Enumeration (GFE) equivalent, while specifically looking at certain gene features (i.e. exon 1, intron 1, all introns, UTRS), where all other features are "zeroed-out". Gene Feature Enumeration (GFE) is a credible bioinformatics tool that was developed to present HLA nomenclature in the context of core and non-core features by enumerating unique nucleotide sequences for each GF in HLA, which allows for quick dissection on which enumerations may be shared in control or disease states amongst alleles. The main wrapper function allows users to specify which "map", or gene feature they would like to look at. Setting info=T allows users to view possible "map" options. After conversion to its GFE equivalent, the dataframe is analyzed in BIGDAWG for statistical processing (i.e. outputs of kx2 and chi-squared tests).

Module 2 extracts variant amino acid positions for a given HLA locus, and analyzes them in BIGDAWG. Based on significant single variant amino acid positions, motifs are built (i.e. pairs, triplets, quadruplets...etc) and move on to the next iteration based on whether they show improvement compared to their previous counterparts ORs (in this case, the OR threshold is +/- 0.1). The function runs until the maximal OR is obtained, or no more improvement is made. Module 2 was programmed based off IPD-IMGT/HLA Release 3.35.0 protein alignments, where the highest motifs produced were sextets -- the program produces motifs up to a maximal OR of sextets. Future improvements will be made to increase this threshold, as well as allow for a user-defined motif list other than step-wise increments (i.e. 2,4,6,8 instead of 1,2,3,4,5). 

